cpp/v2/test/api_test_v2.cc

```diff
@@ -0,0 +1,262 @@
+#include <api_v2.h>
+#include <gtest/gtest.h>
+
+#include <iostream>
+
+using namespace ray;
+
+void hello() { std::cout << "hello\n"; }
+
+void dummy() { std::cout << "dummy\n"; }
+
+int bar(int a) {
+  std::cout << "bar"
+            << ", arg: " << a << "\n";
+  return a;
+}
+
+namespace my_namespace {
+int bar(int a) {
+  std::cout << "bar in namespace"
+            << ", arg: " << a << "\n";
+  return a + 1;
+}
+}  // namespace my_namespace
+
+int add(int a, int b) { return a + b; }
+
+void not_registered_func() {}
+
+int overload_func() {
+  std::cout << "overload_func no arguments\n";
+  return 0;
+}
+int overload_func(int i) {
+  std::cout << "overload_func one argument\n";
+  return i;
+}
+int overload_func(int i, int j) {
+  std::cout << "overload_func two arguments\n";
+  return i + j;
+}
+
+struct Base {
+  Base() = default;
+  Base(int) {}
+  Base(int, ObjectRef<int>) {}
+  ~Base() = default;
+
+  static Base Create(int) { return {}; }
+
+  virtual int bar(int i) {
+    std::cout << "bar in Base\n";
+    return i;
+  }
+  int foo(int i) {
+    std::cout << "foo\n";
+    return i;
+  }
+
+  int overload_func(int i) {
+    std::cout << "Base::overload_func one argument\n";
+    return i;
+  }
+  int overload_func(int i) const {
+    std::cout << "Base::overload_func one argument\n";
+    return i;
+  }
+  int overload_func(int i, int j) {
+    std::cout << "Base::overload_func two arguments\n";
+    return i + j;
+  }
+  int overload_func(int i, int j, int k) {
+    std::cout << "Base::overload_func two arguments\n";
+    return i + j + k;
+  }
+};
+
+struct Derived : public Base {
+  int bar(int i) override {
+    std::cout << "bar in Derived\n";
+    return i + 1;
+  }
+};
+
+RAY_REGISTER(hello, bar, add, my_namespace::bar, &Base::foo, &Base::bar,
+             RayFunc(overload_func), RayFunc(overload_func, int),
+             RayMemberFunc(&Base::overload_func, cv_none, int),
```

Comment:
what is `cv_none`? If it means something, can we name it better?

Reply:
C++ member function is very different from free function, it has many qualifiers, such as const, const reference, reference, volatile, const volatile etc.
So we need a tag to show whether it is a non const or const reference or const volatile member function.

There such tags in overload.h to cover all qualifiers situation:

struct cv_none {}; //no cv qualifiers
struct cv_const {}; //only const
struct cv_volatile {};//only volatile
struct cv_cv {}; //const volatile
struct ref_none {};//no reference
struct ref_rval {};//right value reference
struct ref_lval {};//left value reference